File: frontend/src/stores/workflow.ts
import { type NodeData } from '../types/nodes'
import { type Link } from '../types/link'

// Workflow interface
export interface Workflow {
  name: string | null
  workspace: string | null
  workers: number
  nodes: NodeData[]
  links: Link[]
  lastNode: number
  lastLink: number
}

export function defaultWorkflow(workspace?: string | null) {
  return {
    name: null,
    workspace: workspace || null,
    workers: 2,
    nodes: [],
    links: [],
    lastNode: 0,
    lastLink: 0,
  }
}

export function copyWorkflow(workflow: Workflow, newName?: string) {
  const newWorkflow = JSON.parse(JSON.stringify(workflow))
  if (newName) {
    newWorkflow.name = newName
  }
  return newWorkflow
}

File: python/atflow/workflow.py
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field

class WorkflowRun(BaseModel):
	runs: List[int] = Field(default_factory=list)
	tags: List[str] = Field(default_factory=list)

class WorkflowNode(BaseModel):
	id: int
	name: str
	position: Dict[str, float]
	inputs: List[Dict[str, Any]] = Field(default_factory=list)
	outputs: List[Dict[str, Any]] = Field(default_factory=list)
	properties: List[Dict[str, Any]] = Field(default_factory=list)

class WorkflowLink(BaseModel):
	id: int
	source: int
	sourceHandle: str
	target: int
	targetHandle: str

class Workflow(BaseModel):
	name: str
	workspace: Optional[str] = None
	workers: int = 2
	run: WorkflowRun
	nodes: List[WorkflowNode] = Field(default_factory=list)
	links: List[WorkflowLink] = Field(default_factory=list)
	last_node: int = Field(alias="lastNode")
	last_link: int = Field(alias="lastLink")